File: me_daily/2025/10/1028/cf2128e/cf2128e.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var reader = bufio.NewReader(os.Stdin)
var writer = bufio.NewWriter(os.Stdout)

func readLine() string {
	s, _ := reader.ReadString('\n')
	return strings.TrimSpace(s)
}

func readInts() []int {
	fields := strings.Fields(readLine())
	res := make([]int, len(fields))
	for i, f := range fields {
		res[i], _ = strconv.Atoi(f)
	}
	return res
}
type Result struct {
	left, right int
}

func check(vi, n int, a []int, k int) Result {
	s := make([]int, n+1)
	res := Result{0, 0}
	
	for i := 1; i <= n; i++ {
		val := -1
		if a[i] >= vi {
			val = 1
		}
		s[i] = s[i-1] + val
	}
	
	pos := 0
	for i := k; i <= n; i++ {
		if s[i-k] < s[pos] {
			pos = i - k
		}
		if s[i]-s[pos] >= 0 {
			res.left = pos + 1
			res.right = i
			return res
		}
	}
	return res
}

func solve(a []int, n, k int) {
	lo, hi, ans := 1, n, 0
	finalRes := Result{0, 0}
	
	for lo <= hi {
		vi := lo + (hi-lo)/2
		lr := check(vi, n, a, k)
		if lr.left > 0 {
			ans = vi
			finalRes = lr
			lo = vi + 1
		} else {
			hi = vi - 1
		}
	}
	fmt.Printf("%d %d %d\n", ans, finalRes.left, finalRes.right)
}
func main() {
	defer writer.Flush() //main() 结束时自动调用
	t, _ := strconv.Atoi(readLine())
	for ; t > 0; t-- {
		header := readInts()
		n, k := header[0], header[1]
		ar := readInts()
		a := make([]int, n+1)
		for i := 1; i <= n; i++ {
			a[i] = ar[i-1]
		}
		solve(a, n, k)
	}
}

File: me_daily/2025/10/1028/cf2128e/cf2128e.py
from collections import defaultdict
from typing import List
import sys
from collections import Counter
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()

t = II()
for _ in range(t):
    n, k = LII()
    a = [0] + LII()  # 哨兵 a[0]=0
    def check(v):
        sum_ = [0] * (n + 1) 
        for i in range(1, n + 1):
            sum_[i] = 1 if a[i] >= v else -1
            sum_[i] += sum_[i - 1]
        pos = 0
        for i in range(k, n + 1):
            if sum_[i - k] < sum_[pos]:
                pos = i - k
            if sum_[i] - sum_[pos] >= 0:
                return [pos + 1, i]
        return [0, 0]
    
    lo, hi = 1, n
    ans = 0 
    res = [0, 0]  
    while lo <= hi:
        vi = lo + (hi - lo) // 2
        pr = check(vi) 
        if pr[0] > 0:  # 可行，记录并找更大
            ans = vi
            res = pr
            lo = vi + 1
        else:
            hi = vi - 1
    print(ans, res[0], res[1])